### PR TITLE
test with makefile

### PR DESCRIPTION
**Pull request by deusebio**
In spark we are currently having a bit of a more complicated workflow to build images, that goes also through using Dockerfiles to patch ROCK images. 

In in this PR:
* I have been experimenting and getting a bit of a cleaner UX using Makefile
* I have also change the CI/CD pipeline with some safety checks to make sure that branch and tags to be published indeed correspond. This should also allow to use the same workflow for publishing to different tracks and risks (that are consistent with the branch naming)

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -23,11 +25,32 @@ jobs:
...
+
+      - name: Install dependencies
+        run: |
+          sudo snap install docker
+          sudo addgroup --system docker; sudo adduser $USER docker
+          newgrp docker
+          sudo snap disable docker; sudo snap enable docker
+
+          sudo snap install yq
+          sudo snap install rockcraft --classic --edge
```
Comment (grobbie) [inline .github/workflows/build.yaml:40]: Why edge, out of interest? Maybe choose something other than latest?

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -23,11 +25,32 @@ jobs:
...
+      - name: Install dependencies
+        run: |
+          sudo snap install docker
+          sudo addgroup --system docker; sudo adduser $USER docker
+          newgrp docker
+          sudo snap disable docker; sudo snap enable docker
+
+          sudo snap install yq
+          sudo snap install rockcraft --classic --edge    
+          sudo snap install --devmode --channel edge skopeo
```
Comment (grobbie) [inline .github/workflows/build.yaml:41]: Same remark here

File: .github/workflows/integration.yaml [hunk 3]
```diff
@@ -21,17 +21,35 @@ jobs:
...
+
+      - name: Get Artifact Name
+        id: artifact
+        run: |
+          ARTIFACT=$(make help | grep 'Artifact: ')
+          echo "name=${ARTIFACT#'Artifact: '}" >> $GITHUB_OUTPUT
+
+      - name: Install and configure microk8s
+        run: |
+          make microk8s
```
Comment (grobbie) [inline .github/workflows/integration.yaml:33]: Seems like a potential use case for GitHub actions

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,106 @@
+# Makefile macros (or variables) are defined a little bit differently than traditional bash, keep in mind that in the Makefile there's top-level Makefile-only syntax, and everything else is bash script syntax.
```
Comment (grobbie) [inline Makefile:4]: Add the copyright and license header

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+# ======================
+# INTERNAL VARIABLES
+# ======================
+
+_MAKE_DIR := .make_cache
+$(shell mkdir -p $(_MAKE_DIR))
+
+K8S_TAG := $(_MAKE_DIR)/.k8s_tag
+
+PLATFORM=amd64
```
Comment (grobbie) [inline Makefile]: We should make this user definable and allow arm and riscv in future

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+          docker-daemon:"$(_TMP_OCI_NAME):$(VERSION)"
+	if [ ! -d "$(_MAKE_DIR)/$(_TMP_OCI_NAME)" ]; then mkdir -p "$(_MAKE_DIR)/$(_TMP_OCI_NAME)"; fi
+	touch $(_TMP_OCI_TAG)
+
+$(CHARMED_OCI_TAG): $(_TMP_OCI_TAG)
+	docker build - -t "$(CHARMED_OCI_FULL_NAME):$(VERSION)" --build-arg BASE_IMAGE="$(_TMP_OCI_NAME):$(VERSION)" < Dockerfile
+	if [ ! -d "$(_MAKE_DIR)/$(CHARMED_OCI_FULL_NAME)" ]; then mkdir -p "$(_MAKE_DIR)/$(CHARMED_OCI_FULL_NAME)"; fi
+	touch $(CHARMED_OCI_TAG)
+
+$(K8S_TAG):
```
Comment (grobbie) [inline Makefile:75]: This could do with some comments to explain what's going on. Same for `microk8s:` below

**Review by grobbie — CHANGES_REQUESTED** [review]
A couple of things but otherwise good work

File: .github/workflows/build.yaml (hunk 1 above)
Comment (deusebio) [reply .github/workflows/build.yaml:40]: I believe only `edge` exists actually! :D

File: .github/workflows/build.yaml (hunk 2 above)
Comment (deusebio) [reply .github/workflows/build.yaml:41]: Also here, only `edge` exists

File: .github/workflows/integration.yaml (hunk 3 above)
Comment (deusebio) [reply .github/workflows/integration.yaml:33]: Uhm, that's interesting, actually, and indeed there is an action that may do this (https://github.com/balchua/microk8s-actions). But one of the reason to keep using make is that `make microk8s` command can be used also locally (therefore out of github) to setup microk8s (and related services needed by spark). I find this quite convinient when setting up locally the environment to be able to run integration tests

File: Makefile (hunk 5 above)
Comment (deusebio) [reply Makefile]: This should be user definable already. Variables defined in Makefile can be easily overridden, e.g. `make build PLATFORM=arm`, although now we miss the building logic when we change this. But we should address this in the future. Anyhow, I have now also moved its position in the file and place it together with the other External variables.

File: Makefile (hunk 6 above)
Comment (deusebio) [reply Makefile:75]: I have added a comment in the help referring to the microk8s rule, and also placed the same comments in the different rules for better clarity

**Comment by deusebio** [comment]
Not relevant anymore

**Pull request closed without merging**